pkg/ensurance/cm/advanced_cpu_manager.go

```diff
@@ -0,0 +1,354 @@
+}
+
+func (m *AdvancedCpuManager) reconcileState() {
+	m.syncState(true)
+	sharedCPUSet := m.getSharedCpu().Union(m.state.GetDefaultCPUSet())
+	for _, pod := range m.activepods() {
+		for _, container := range pod.Spec.Containers {
+			containerID := GetContainerIdFromPod(pod, container.Name)
+			cset, ok := m.state.GetCPUSet(string(pod.UID), container.Name)
+			if !ok {
+				cset = sharedCPUSet
+			}
+			klog.Infof("[Advancedcpumanager] reconcileState: updating container (pod: %s, container: %s, container id: %s, cpuset: \"%v\")",
+				pod.Name, container.Name, containerID, cset)
+			err := m.updateContainerCPUSet(containerID, cset)
+			if err != nil {
+				klog.Errorf("[Advancedcpumanager] reconcileState: failed to update container (pod: %s, container: %s, container id: %s, cpuset: \"%v\", error: %v)",
+					pod.Name, container.Name, containerID, cset, err)
+				continue
+			}
+		}
+	}
+}
+
+func (m *AdvancedCpuManager) syncState(doAllocate bool) {
+	// We grab the lock to ensure that no new containers will grab CPUs while
+	// executing the code below. Without this lock, its possible that we end up
+	// removing state that is newly added by an asynchronous call to
+	// AddContainer() during the execution of this code.
+	m.Lock()
+	defer m.Unlock()
+	assignments := m.state.GetCPUAssignments()
+
+	// Build a list of (podUID, containerName) pairs for all need to be assigned containers in all active Pods.
+	toBeAssignedContainers := make(map[string]map[string]struct{})
+	for _, pod := range m.activepods() {
+		toBeAssignedContainers[string(pod.UID)] = make(map[string]struct{})
+		for _, container := range pod.Spec.Containers {
+			if m.policy.NeedAllocated(pod, &container) {
+				toBeAssignedContainers[string(pod.UID)][container.Name] = struct{}{}
+				if _, ok := assignments[string(pod.UID)][container.Name]; !ok && doAllocate {
+					err := m.policy.Allocate(m.state, pod, &container)
+					if err != nil {
+						klog.Errorf("[Advancedcpumanager] Allocate error: %v", err)
+						continue
+					}
+				}
+			}
+		}
+	}
+
+	// Loop through the CPUManager state. Remove any state for containers not
+	// in the `toBeAssignedContainers` list built above.
+	for podUID := range assignments {
+		for containerName := range assignments[podUID] {
+			if _, ok := toBeAssignedContainers[podUID][containerName]; !ok {
+				klog.Errorf("[Advancedcpumanager] removeStaleState: removing (pod %s, container: %s)", podUID, containerName)
+				err := m.policyRemoveContainerByRef(podUID, containerName)
+				if err != nil {
+					klog.Errorf("[Advancedcpumanager] removeStaleState: failed to remove (pod %s, container %s), error: %v)", podUID, containerName, err)
+				}
+			}
+		}
+	}
+}
+
+func (m *AdvancedCpuManager) policyRemoveContainerByRef(podUID string, containerName string) error {
+	err := m.policy.RemoveContainer(m.state, podUID, containerName)
+	return err
+}
+
+func (m *AdvancedCpuManager) updateContainerCPUSet(containerID string, cpus cpuset.CPUSet) error {
```

Comment:
Do you tested for change the cpuset when original is 1,3,5, then it changed to 3,5,6,8  I am not sure if this is ok.  Does it modify the cpuset successfully. 

And it is ok when modify the cpuset, will it update the default other cgroup params by default.

Reply:
1.It's ok to change cgroup when pod is running.
2.if cpus are allocated, `sharedCPUSet := m.getSharedCpu().Union(m.state.GetDefaultCPUSet())`  will be changed,then update other cgroup in loop